### PR TITLE
V3.0.0

### PR DESCRIPTION
**Pull request by fed135**
## Summary

Version 3 proposition

## Breaking Changes!

- Changes the Multi-item query interface, separated from the single-item one. Multi-syntax `getMany` uses itemized Promise.allSettled formatted responses.
- Changed API wording
  - `batch.tick` setting becomes `batch.delay`
  - `batch.max` setting becomes `batch.limit`
  - `uniqueParams` setting becomes `delimiter`
- Removed custom response parser, which felt redundant. Now only itemized objects will be accepted

## Other changes 

- Fixed TS definitions
- Updated libraries
- Updated linting
- Added Github actions to replace Travis

## Merge Checklist

 - [x] Verified locally
 - [x] Unit Tests added
 - [x] Integration tests added
 - [x] Unit tests pass locally
 - [x] Integration tests pass locally
 - [x] Benchmark tests pass locally

### CODE REVIEW

File: src/index.js [hunk 1]
```diff
@@ -38,72 +28,52 @@ class HaStore extends EventEmitter {
...
-   * @param {object} params (Optional)The Request parameters
-   * @returns {Promise} The eventual single record
-   */
-  async get(ids, params = {}, agg = null) {
+  get(ids, params = {}, agg = null) {
+    if (params === null) params = {};
+    const key = contextKey(this.config.delimiter, params);
+    return this.queue.getHandles(key, [ids], params, agg)
+      .then(handles => handles[0]);
+  }
```
Comment (mats852) [inline src/index.js:40]: [⏳] Singular word `id` helps undestanding the context.

Maybe, if it's an array, we could shift the array and grab the first value for extra safety.

```suggestion
  get(id, params = {}, agg = null) {
    if (params === null) params = {};
    const key = contextKey(this.config.delimiter, params);
    return this.queue.getHandles(key, [id], params, agg)
      .then(handles => handles[0]);
  }
```

File: src/index.js [hunk 2]
```diff
@@ -38,72 +28,52 @@ class HaStore extends EventEmitter {
...
-    const key = contextKey(this.config.uniqueParams, params);
-    const handles = await this.queue.getHandles(key, requestIds, params, agg);
-    return Promise.all(handles)
-      .then(response => (!Array.isArray(ids)) ? response[0] : response);
+    const key = contextKey(this.config.delimiter, params);
+    return this.queue.getHandles(key, ids, params, agg)
+      .then((handles) => Promise.allSettled(handles)
+        .then((responses) => ids.reduce((acc, curr, index) => {
+          acc[curr] = responses[index];
+          return acc;
```
Comment (mats852) [inline src/index.js:49]: [⚪] Context in variable names
```suggestion
        .then((outcomes) => ids.reduce((handles, id, index) => {
          handles[id] = outcomes[index];
          return handles;
```

File: src/queries.js [hunk 3]
```diff
@@ -1,109 +1,108 @@
...
-        let numMisses = 0;
+    let numCoalesced = 0;
+    let numCached = 0;
+    let numMisses = 0;
 
-        const handles = [];
-        for (let i = 0; i < ids.length; i++) {
-            handles[i] = undefined;
+    const handles = [];
+    for (let i = 0; i < ids.length; i++) {
```
Comment (mats852) [inline src/queries.js:14]: [💨] It's a matter of style, but since we don't mutate the `i` index, we could use const.

File: src/queries.js [hunk 4]
```diff
@@ -1,109 +1,108 @@
...
 
-    function size() {
-        const contexts = Object.keys(queries);
-        return {
-            contexts: contexts.length,
-            queries: contexts.reduce((acc, curr) => acc + queries[curr].length, 0),
-        }
+    if (numCached > 0) emitter.emit('cacheHit', { key, found: numCached });
+    if (numMisses > 0) emitter.emit('cacheMiss', { key, found: numMisses });
+    if (numCoalesced > 0)  emitter.emit('coalescedHit', { key, found: numCoalesced });
```
Comment (mats852) [inline src/queries.js:43]: [💨] Extra space
```suggestion
    if (numCoalesced > 0) emitter.emit('coalescedHit', { key, found: numCoalesced });
```

File: src/queries.js (hunk 3 above)
Comment (fed135) [reply src/queries.js:14]: Because of the i++ in the loop, I don't think a const would work

**Review by drawm — APPROVED** [review]
Great config & api changes, LGTM!

File: src/queries.js (hunk 3 above)
Comment (mats852) [reply src/queries.js:14]: Totally right!

**Pull request merged**